plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -509,6 +515,58 @@ inferTypeM (Error ann ty) = do
     checkKindM ann ty $ Type ()
     normalizeTypeM $ void ty
 
+-- resTy ~> vResTy     vResTy = sop s_0 ... s_i ... s_n     s_i = [p_0 ... p_m]
+-- [check| G !- t_j : p_j]
+-- ----------------------------------------------------------------------------
+-- [infer| G !- constr resTy i t_0 ... t_n : vResTy]
+inferTypeM t@(Constr ann resTy i args) = do
+    vResTy <- normalizeTypeM $ void resTy
+
+    -- We don't know exactly what to expect, we only know what the i-th sum should look like, so we
+    -- assert that we should have some types in the sum up to there, and then the known product type.
+    let expectedType = TySOP () (replicate ((fromIntegral i) - 1) [dummyType] ++ [replicate (length args) dummyType])
```

Comment:
I don't get the comment. They all seem to be `dummyType` in there?

Reply:
It's unclear. I mean we know the _shape_ of one of the product types, not the whole thing.